src/lib.rs

```diff
@@ -164,25 +169,45 @@ pub fn initialize_check_rotation(device_context: &DeviceContext) -> Result<InitA
         .block_on(initialize_check_rotation_async(device_context))
 }
 
+fn form_rotation_needed(
+    rotations_needed: EitherOrBoth<UserId, Vec1<GroupId>>,
+    ironoxide: IronOxide,
+) -> InitAndRotationCheck {
+    InitAndRotationCheck::RotationNeeded(
+        ironoxide,
+        PrivateKeyRotationCheckResult { rotations_needed },
+    )
+}
+
 async fn initialize_check_rotation_async(
     device_context: &DeviceContext,
 ) -> Result<InitAndRotationCheck> {
-    internal::user_api::user_get_current(device_context.auth())
-        .await
-        .and_then(|curr_user| {
-            let ironoxide = IronOxide::create(&curr_user, &device_context);
-
-            if curr_user.needs_rotation() {
-                Ok(InitAndRotationCheck::RotationNeeded(
-                    ironoxide,
-                    PrivateKeyRotationCheckResult {
-                        rotations_needed: EitherOrBoth::Left(curr_user.account_id().clone()),
-                    },
-                ))
-            } else {
-                Ok(InitAndRotationCheck::NoRotationNeeded(ironoxide))
-            }
-        })
+    use EitherOrBoth::{Both, Left, Right};
+    let (curr_user, group_list_result) = futures::try_join!(
+        internal::user_api::user_get_current(device_context.auth()),
+        internal::group_api::list(device_context.auth(), None)
+    )?;
+    let account_id = curr_user.account_id().to_owned();
+    let ironoxide = IronOxide::create(&curr_user, &device_context);
+    let user_groups = group_list_result.result();
+
+    let groups_needing_rotation = user_groups
+        .into_iter()
+        .filter(|meta_result| meta_result.needs_rotation() == Some(true))
+        .map(|meta_result| meta_result.id().to_owned())
+        .collect::<Vec<_>>();
+
+    // If this is a Some, there are groups needing rotation
+    let maybe_groups_needing_rotation = Vec1::try_from_vec(groups_needing_rotation).ok();
+
+    Ok(
+        match (curr_user.needs_rotation(), maybe_groups_needing_rotation) {
+            (false, None) => InitAndRotationCheck::NoRotationNeeded(ironoxide),
+            (true, None) => form_rotation_needed(Left(account_id), ironoxide),
+            (false, Some(groups)) => form_rotation_needed(Right(groups), ironoxide),
+            (true, Some(groups)) => form_rotation_needed(Both(account_id, groups), ironoxide),
+        },
+    )
```

Comment:
I think this is a good candidate for a private function that can be unit tested.

Reply:
How would you break this up so that the private function doesn't need an ironoxide, so that I can write a unit test on it?

Reply:
I needed to make some things pub(crate) so that I could build them from scratch for the test, but I was able to add a unit test for this function.